wfcGen.py

```diff
@@ -2,83 +2,79 @@
-    # u will newer need that
+        # [left, top, right, bottom]uwu
+        # u will newer need that
 
-    for y in range(lcy):
-        for x in range(lcx):
-            flag = -1
-            for key in cells.keys():
-                if cropped[ins][y][x] == cells[key]["img"]:
-                    flag = key
+        for y in range(lcy):
+            for x in range(lcx):
+                cellId = -1
+                for CellIdKey in cells.keys():
+                    if cropped[ins][y][x] == cells[CellIdKey]["img"]:
+                        cellId = CellIdKey
 
-            if flag != -1:
-                cells[flag]["nbg"][0].append(cropped[ins][y][x-1-(res-1)*int(cropShift)])
-                cells[flag]["nbg"][1].append(cropped[ins][y-1-(res-1)*int(cropShift)][x])
-                cells[flag]["nbg"][2].append(cropped[ins][y][(x+1+(res-1)*int(cropShift))%lcx])
-                cells[flag]["nbg"][3].append(cropped[ins][(y+1+(res-1)*int(cropShift))%lcy][x])
-            else:
-                cells[str(len(cells.keys()))] = {
-                    "nbg" : {
-                        "0" : [cropped[ins][y][x-1-(res-1)*int(cropShift)]],
-                        "1" : [cropped[ins][y-1-(res-1)*int(cropShift)][x]],
-                        "2" : [cropped[ins][y][(x+1+(res-1)*int(cropShift))%lcx]],
-                        "3" : [cropped[ins][(y+1+(res-1)*int(cropShift))%lcy][x]],
-                    }, 
-                    "img" : cropped[ins][y][x]
-                }
-
-    for y in range(lcy):
-        for x in range(lcx):
-            for key in cells.keys():
-                if cells[key]["img"] == cropped[ins][y][x]:
-                    Gkey = key
-            for key in cells.keys():
-                for key2 in cells[key]["nbg"].keys():
-                    if key2 != "img":
-                        for i in range(len(cells[key]["nbg"][key2])):
-                            if cells[key]["nbg"][key2][i] == cropped[ins][y][x]:
-                                cells[key]["nbg"][key2][i] = Gkey
-
-    print(cells)
+                if cellId != -1:
+                    cells[cellId]["nbg"][0].append(cropped[ins][y][x-1-(tileSize-1)*int(cropShift)])
+                    cells[cellId]["nbg"][1].append(cropped[ins][y-1-(tileSize-1)*int(cropShift)][x])
+                    cells[cellId]["nbg"][2].append(cropped[ins][y][(x+1+(tileSize-1)*int(cropShift))%lcx])
+                    cells[cellId]["nbg"][3].append(cropped[ins][(y+1+(tileSize-1)*int(cropShift))%lcy][x])
+                else:
+                    cells[len(cells.keys())] = {
+                        "nbg" : {
+                            0 : [cropped[ins][y][x-1-(tileSize-1)*int(cropShift)]],
+                            1 : [cropped[ins][y-1-(tileSize-1)*int(cropShift)][x]],
+                            2 : [cropped[ins][y][(x+1+(tileSize-1)*int(cropShift))%lcx]],
+                            3 : [cropped[ins][(y+1+(tileSize-1)*int(cropShift))%lcy][x]],

```

Comment:
I suggest that rather than using indexing, you use names for the directions in which the cells are adjacent. I think a `NamedTuple` fits well here

Reply:
Have a look at [this SO question](https://stackoverflow.com/q/42618134/11248508). One of the answers gives an example of how one can work with `NamedTuple`s if they want to change their values, as well as suggests an alternative which allows for mutable values similar to `NamedTuple` in how you access them